pydomcfg/domzgr/zco.py

```diff
@@ -0,0 +1,293 @@
+            (default = pp_to_be_computed, i.e. computed from
+            ppdzmin, pphmax, ppkth and ppacr)
+        ppa0: float
+            Coeff. controlling distibution of vertical levels
+            (default = pp_to_be_computed, i.e. computed from
+            ppdzmin, pphmax, ppkth and ppacr)
+        ppa1: float
+            Coeff. controlling distibution of vertical levels
+            (default = pp_to_be_computed, i.e. computed from
+            ppdzmin, pphmax, ppkth and ppacr)
+        ldbletanh: bool
+            Logical flag to use or not double tanh stretching function
+            (default = False)
+        ppa2: float
+            Double tanh stretching function parameter
+        ppkth2: float
+            Double tanh stretching function parameter
+        ppacr2: float
+            Double tanh stretching function parameter
+        ln_e3_dep: bool
+            Logical flag to comp. e3 as fin. diff. (True) or
+            analyt. (False) (default = True)
+
+        Returns
+        -------
+        Dataset
+           Describing the 3D geometry of the model
+
+        Raises
+        -------
+        ValueError
+            If ldbletanh = True and parametrs are equal to 0.
+
+        """
+        self._ppdzmin = ppdzmin
+        self._pphmax = pphmax
+        self._ppkth = ppkth
+        self._ppacr = ppacr
+        self._ppsur = ppsur
+        self._ppa0 = ppa0
+        self._ppa1 = ppa1
+        self._ldbletanh = ldbletanh
+        self._ppa2 = ppa2
+        self._ppkth2 = ppkth2
+        self._ppacr2 = ppacr2
+        self._is_uniform = not ppkth or not ppacr
+        self._ln_e3_dep = ln_e3_dep
+
+        # Checking consistency of input parameters
+        if self._ldbletanh:
+            if self._ppa2 * self._ppkth2 * self._ppacr2 == 0.0:
+                raise ValueError(
+                    "ppa2, ppkth2 and ppacr2 MUST be > 0. " "when ldbletanh = True"
+                )
+
+        ds = self._init_ds()
+
+        # computing coeff. if needed
+        self._ppsur, self._ppa0, self._ppa1 = self._compute_pp()
+
+        # compute sigma-coordinates for z3 computation
+        kindx = ds["z"]
+        sigma, sigma_p1 = (self._compute_sigma(kk) for kk in (kindx, kindx + 1.0))
+        self._sigT, self._sigW = sigma
+        self._sigTp1, self._sigWp1 = sigma_p1
+
+        # compute z3 depths of zco vertical levels
+        dsz = self._zco_z3(ds)
+
+        # compute e3 scale factors
+        dse = self._compute_e3(dsz) if self._ln_e3_dep else self._analyt_e3(dsz)
+
+        return dse
+
+    # --------------------------------------------------------------------------
+    def _compute_pp(self) -> tuple:
+        """
+        Compute the coefficients for zco grid if needed.
+
+        Returns
+        -------
+        tuple
+            (ppsur, ppa0, ppa1)
+
+        """
+        pp_in = (self._ppsur, self._ppa0, self._ppa1)
+        if not pp_in.count(self.pp_to_be_computed):
+            return pp_in
```

Comment:
```suggestion
        if all(is_nemo_none(pp) for pp in pp_in):
            return pp_in
```

Reply:
Implementing this is creating problem with bloody mypy ... any tips?

pydomcfg/domzgr/zco.py:232: error: Incompatible types in assignment (expression has type "Optional[float]", variable has type "float")  [assignment]
pydomcfg/domzgr/zco.py:233: error: Unsupported operand types for * ("None" and "int")  [operator]
pydomcfg/domzgr/zco.py:233: note: Left operand is of type "Optional[float]"
pydomcfg/domzgr/zco.py:234: error: Unsupported operand types for * ("None" and "int")  [operator]
pydomcfg/domzgr/zco.py:234: note: Left operand is of type "Optional[float]"

where the mentioned lines are just:

a1 = self._ppsur
a2 = self._ppa0 * (self._jpk - 1)
a3 = self._ppa1 * self._ppacr